File: 844. Backspace string compare/backspace_string_compare.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
 var backspaceCompare = function(s, t) {

    return newEdits(s) == newEdits(t);

    // create a new function, to get the "counter" to store the skipping counts

    function newEdits(str){
        let result = ''; // create an empty string
        let backspaces = 0;

        for(let i = str.length - 1 ; i >= 0; i -=1){
            if(str[i] === '#'){
                backspaces += 1;
            }else if(backspaces > 0){
                backspaces -= 1;
            }else{
                result = str[i] + result;
            }
        }
        return result;
    }
 };

File: 844. Backspace string compare/backspace_string_compare.py
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        # for this problem, we could look at the strings backwards; if encoutering a "#", the next lowercase char will be skipped.
        # and we initialize two variables to store the skipping counts

        i = len(s) - 1
        j = len(t) - 1

        backS = backT = 0 # variables to store the skipping counts

        while True: # both s and t are not None
            while i >= 0 and (backS != 0 or s[i] == '#'):
                backS += 1 if s[i] == '#' else -1
                i -= 1
            while j >= 0 and (backT != 0 or t[j] == '#'):
                backT += 1 if t[j] == '#' else -1
                j -= 1

            if not (i >= 0 and j >= 0 and s[i] == t[j]): # if not the abve two cases, both i and j are -1
                return i == j == -1 

            i , j = i - 1, j - 1

